File: app/src/main/java/com/juniperphoton/myerlist/model/ToDoCategory.java
package com.juniperphoton.myerlist.model;


import android.graphics.Color;
import android.support.v4.content.ContextCompat;

import com.juniperphoton.myerlist.App;
import com.juniperphoton.myerlist.R;
import com.juniperphoton.myerlist.util.ColorUtil;
import com.juniperphoton.myerlist.util.LocalSettingUtil;
import com.juniperphoton.myerlist.util.Params;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class ToDoCategory extends RealmObject {
    public static final String ID_KEY = "id";
    public static final String POSITION_KEY = "position";
    public static final String SID_KEY = "sid";

    public static final int ALL_ID = 0;
    public static final int DELETED_ID = -1;
    public static final int PERSONALIZATION_ID = -2;

    private String name;
    private String color;

    private int position;

    @PrimaryKey
    private int id;

    private String sid;

    public ToDoCategory() {

    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getColor() {
        return color;
    }

    public int getIntColor() {
        return Color.parseColor(color);
    }

    public void setColor(String color) {
        this.color = color;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public void setSid(String sid) {
        this.sid = sid;
    }

    public static ToDoCategory getAllCategory() {
        ToDoCategory category = new ToDoCategory();
        category.setColor(ColorUtil.fromInt(ContextCompat.getColor(App.getInstance(), R.color.MyerListBlue)));
        category.setName(App.getInstance().getString(R.string.all));
        category.setId(ALL_ID);
        category.setSid(LocalSettingUtil.getString(App.getInstance(), Params.SID_KEY));
        return category;
    }

    public static ToDoCategory getDeletedCategory() {
        ToDoCategory category = new ToDoCategory();
        category.setColor(ColorUtil.fromInt(ContextCompat.getColor(App.getInstance(), R.color.DeleteColor)));
        category.setName(App.getInstance().getString(R.string.deleted));
        category.setId(DELETED_ID);
        return category;
    }

    public static ToDoCategory getPersonalizationCategory() {
        ToDoCategory category = new ToDoCategory();
        category.setColor(ColorUtil.fromInt(Color.WHITE));
        category.setName(App.getInstance().getString(R.string.personalization));
        category.setId(PERSONALIZATION_ID);
        return category;
    }

    public ToDoCategory getCopy() {
        ToDoCategory category = new ToDoCategory();
        category.name = name;
        category.id = id;
        category.color = color;
        category.position = position;
        return category;
    }

    public int getPosition() {
        return position;
    }

    public void setPosition(int position) {
        this.position = position;
    }
}

File: app/src/main/java/com/juniperphoton/myerlist/model/ToDoCategory.kt
package com.juniperphoton.myerlist.model

import android.graphics.Color
import com.juniperphoton.myerlist.App
import com.juniperphoton.myerlist.R
import com.juniperphoton.myerlist.extension.getResColor
import com.juniperphoton.myerlist.extension.getResString
import com.juniperphoton.myerlist.extension.toColorString
import com.juniperphoton.myerlist.util.LocalSettingUtil
import com.juniperphoton.myerlist.util.Params
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey

open class ToDoCategory : RealmObject() {
    companion object {
        const val KEY_ID = "id"
        const val KEY_POSITION = "position"
        const val KEY_SID = "sid"

        const val VALUE_ALL_ID = 0
        const val VALUE_DELETED_ID = -1
        const val VALUE_PERSONALIZATION_ID = -2

        val allCategory: ToDoCategory
            get() {
                val category = ToDoCategory()
                category.color = R.color.MyerListBlue.getResColor().toColorString()
                category.name = R.string.all.getResString()
                category.id = VALUE_ALL_ID
                category.setSid(LocalSettingUtil.getString(App.instance!!, Params.SID_KEY)!!)
                return category
            }

        val deletedCategory: ToDoCategory
            get() {
                val category = ToDoCategory()
                category.color = R.color.DeleteColor.getResColor().toColorString()
                category.name = R.string.deleted.getResString()
                category.id = VALUE_DELETED_ID
                return category
            }

        val personalizationCategory: ToDoCategory
            get() {
                val category = ToDoCategory()
                category.color = Color.WHITE.toColorString()
                category.name = R.string.personalization.getResString()
                category.id = VALUE_PERSONALIZATION_ID
                return category
            }
    }

    var name: String? = null
    var color: String? = null

    var position: Int = 0

    @PrimaryKey
    var id: Int = 0

    private var sid: String? = null

    val intColor: Int
        get() = Color.parseColor(color)

    fun setSid(sid: String) {
        this.sid = sid
    }

    val copy: ToDoCategory
        get() {
            val category = ToDoCategory()
            category.name = name
            category.id = id
            category.color = color
            category.position = position
            return category
        }
}
